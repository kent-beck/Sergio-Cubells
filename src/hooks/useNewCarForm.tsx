import { useCallback, useReducer } from "react";
import { Car, ChangeValuePayload } from "../types";

type FormReducerAction =
  | {
      type: "change_value";
      payload: ChangeValuePayload;
    }
  | { type: "clear" };

export const CAR_INITIAL_STATE = {
  title: "",
  euroPrice: "",
  image: "",
  description: "",
};

export const getChangedValue = (state: Car, payload: ChangeValuePayload) => {
  const { inputName, inputValue } = payload;
  return {
    ...state,
    [inputName]: inputValue,
  };
};

const formReducer = (state: Car, action: FormReducerAction) => {
  switch (action.type) {
    case "change_value":
      return getChangedValue(state, action.payload);
    case "clear":
      return CAR_INITIAL_STATE;
  }
};

const useNewCarForm = () => {
  const [inputValues, dispatch] = useReducer(formReducer, CAR_INITIAL_STATE);

  const clearForm = useCallback(() => dispatch({ type: "clear" }), []);

  const changeValue = useCallback(
    (name: string, value: string) =>
      dispatch({
        type: "change_value",
        payload: {
          inputName: name,
          inputValue: value,
        },
      }),
    []
  );

  return {
    formState: inputValues,
    clearForm,
    changeValue,
  };
};

export default useNewCarForm;
