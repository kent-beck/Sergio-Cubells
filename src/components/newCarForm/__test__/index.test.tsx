import { fireEvent, render, screen } from "@testing-library/react";
import NewCarForm from "..";
import HeaderDetail from "..";
import { CAR_INITIAL_STATE } from "../../../hooks/useNewCarForm";

const onCloseSpy = jest.fn();
const onNewCar = jest.fn();

const newCarForm = (
  <NewCarForm isVisible={true} onClose={onCloseSpy} onNewCar={onNewCar} />
);

describe("Render a ", () => {
  test("New Car Form", () => {
    const tree = render(newCarForm).asFragment();
    expect(tree).toMatchSnapshot();
  });

  test("empty fragment, form is not visible", () => {
    const tree = render(
      <NewCarForm isVisible={false} onClose={onCloseSpy} onNewCar={onNewCar} />
    ).asFragment();

    expect(tree).toBeInstanceOf(DocumentFragment);
    expect(tree.childNodes.length).toBe(0);
  });

  test("new car text", () => {
    render(newCarForm);

    const newCarTextElement = screen.getByText("Nuevo Coche");

    expect(newCarTextElement).toBeInTheDocument();
  });

  test("save text", () => {
    render(newCarForm);

    const saveTextElement = screen.getByText("Guardar");

    expect(saveTextElement).toBeInTheDocument();
  });

  test(" 3 inputs ", () => {
    render(newCarForm);

    const inputFormList = screen.getAllByTestId("inputForm");

    expect(inputFormList.length).toBe(3);
  });

  test(" a textArea", () => {
    render(newCarForm);

    const textAreaList = screen.getAllByTestId("textAreaForm");

    expect(textAreaList.length).toBe(1);
    expect(textAreaList[0]).toBeInstanceOf(HTMLTextAreaElement);
  });
});

describe("Click on back icon", () => {
  test("close icon called", () => {
    render(newCarForm);

    const closeFormIcon = screen.getByTestId("closeFormIcon");

    fireEvent.click(closeFormIcon);

    expect(onCloseSpy).toBeCalled();
  });

  describe("saveCarButton clicked", () => {
    test("onClose is called", () => {
      render(newCarForm);

      const saveCarButton = screen.getByTestId("saveCarButton");

      fireEvent.click(saveCarButton);

      expect(onCloseSpy).toBeCalled();
    });

    test("onNewCar is called", () => {
      render(newCarForm);

      const saveCarButton = screen.getByTestId("saveCarButton");

      fireEvent.click(saveCarButton);

      expect(onNewCar).toBeCalled();
      expect(onNewCar).toBeCalledWith(CAR_INITIAL_STATE);
    });
  });
});
