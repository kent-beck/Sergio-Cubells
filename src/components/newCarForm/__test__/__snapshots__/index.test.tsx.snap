// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render a  New Car Form 1`] = `
<DocumentFragment>
  <div
    class="modalContainer"
  >
    <form
      class="newCarForm"
    >
      <div
        class="formTitleContainer"
      >
        <div
          class="formTitleText"
        >
          Nuevo Coche
        </div>
        <img
          class="closeFormIcon"
          data-testid="closeFormIcon"
          src="close.png"
        />
      </div>
      <input
        class="inputBorderBottom"
        data-testid="inputForm"
        name="title"
        placeholder="Nombre del coche"
        type="text"
        value=""
      />
      <input
        class="inputBorderBottom"
        data-testid="inputForm"
        name="euroPrice"
        placeholder="Precio en Euros"
        type="text"
        value=""
      />
      <input
        class="inputBorderBottom"
        data-testid="inputForm"
        name="image"
        placeholder="Url de la imagen"
        type="text"
        value=""
      />
      <textarea
        class="inputBorderBottom"
        data-testid="textAreaForm"
        name="description"
        placeholder="DescripciÃ³n"
      />
      <button
        class="addCarButton"
        data-testid="saveCarButton"
        type="submit"
      >
        Guardar
      </button>
    </form>
  </div>
</DocumentFragment>
`;
